sys/riscv/pmap.c

```diff
@@ -1,168 +1,708 @@
+  TAILQ_INIT(&pmap->pte_pages);
+  TAILQ_INIT(&pmap->pv_list);
 }
 
-pmap_t *pmap_lookup(vaddr_t va) {
-  panic("Not implemented!");
+void init_pmap(void) {
+  kernel_pmap.pde = kernel_pde;
+  pmap_setup(&kernel_pmap);
 }
 
-pmap_t *pmap_kernel(void) {
-  panic("Not implemented!");
+pmap_t *pmap_new(void) {
+  pmap_t *pmap = pool_alloc(P_PMAP, M_ZERO);
+  vm_page_t *pg = pmap_pagealloc();
+  pmap->pde = pg->paddr;
+  pmap_setup(pmap);
+
+  /* Install kernel pagetables. */
+  const size_t off = PAGESIZE / 2;
+  WITH_MTX_LOCK (&kernel_pmap.mtx) {
+    memcpy((void *)phys_to_dmap(pmap->pde) + off,
+           (void *)phys_to_dmap(kernel_pde) + off, PAGESIZE / 2);
+  }
+
+  TAILQ_INSERT_TAIL(&pmap->pte_pages, pg, pageq);
+  klog("Page directory table allocated at %p", pmap->pde);
+
+  WITH_MTX_LOCK (&user_pmaps_lock) {
+    LIST_INSERT_HEAD(&user_pmaps, pmap, pmap_link);
+  }
+
+  return pmap;
 }
 
-pmap_t *pmap_user(void) {
-  panic("Not implemented!");
+void pmap_activate(pmap_t *pmap) {
+  SCOPED_NO_PREEMPTION();
+
+  pmap_t *old = PCPU_GET(curpmap);
+  if (pmap == old)
+    return;
+
+  __set_satp(pmap->satp);
+  PCPU_SET(curpmap, pmap);
+
+  __sfence_vma();
+}
+
+void pmap_delete(pmap_t *pmap) {
+  assert(pmap != pmap_kernel());
+
+  WITH_MTX_LOCK (&user_pmaps_lock) { LIST_REMOVE(pmap, pmap_link); }
+
+  while (!TAILQ_EMPTY(&pmap->pv_list)) {
+    pv_entry_t *pv = TAILQ_FIRST(&pmap->pv_list);
+    vm_page_t *pg;
+    paddr_t pa;
+    assert(pmap_extract_nolock(pmap, pv->va, &pa));
+    pg = vm_page_find(pa);
+    WITH_MTX_LOCK (&pv_list_lock) { TAILQ_REMOVE(&pg->pv_list, pv, page_link); }
+    TAILQ_REMOVE(&pmap->pv_list, pv, pmap_link);
+    pool_free(P_PV, pv);
+  }
+
+  while (!TAILQ_EMPTY(&pmap->pte_pages)) {
+    vm_page_t *pg = TAILQ_FIRST(&pmap->pte_pages);
+    TAILQ_REMOVE(&pmap->pte_pages, pg, pageq);
+    vm_page_free(pg);
+  }
+
+  free_asid(pmap->asid);
+  pool_free(P_PMAP, pmap);
+}
+
+void pmap_bootstrap(paddr_t pd_pa, vaddr_t pd_va) {
+  dmap_paddr_base = kenv_get_ulong("mem_start");
+  dmap_paddr_end = kenv_get_ulong("mem_end");
+  size_t dmap_size = dmap_paddr_end - dmap_paddr_base;
+  kernel_pde = pd_pa;
+
+  /* Assume physical memory starts at the beginning of L0 region. */
+  assert(is_aligned(dmap_paddr_base, L0_SIZE));
+
+  /* We must have enough virtual addresses. */
+  assert(dmap_size <= DMAP_MAX_SIZE);
+
+  /* For RV32 we assume 32-bit physical address space. */
+  assert(dmap_paddr_base < dmap_paddr_end);
+
+  klog("Physical memory range: %p - %p", dmap_paddr_base, dmap_paddr_end - 1);
+
+  klog("dmap range: %p - %p", DMAP_VADDR_BASE, DMAP_VADDR_BASE + dmap_size - 1);
+
+  /* Build direct map using superpages. */
+  pd_entry_t *pde = (void *)pd_va;
+  size_t idx = L0_INDEX(DMAP_VADDR_BASE);
+  for (paddr_t pa = dmap_paddr_base; pa < dmap_paddr_end; pa += L0_SIZE, idx++)
+    pde[idx] = PA_TO_PTE(pa) | PTE_KERN;
+
+  __sfence_vma();
+
+  void *fdtp = (void *)phys_to_dmap(FDT_get_physaddr());
+  FDT_changeroot(fdtp);
```

Comment:
Why? `pmap` should be independent module. Here you are trying to do some weird hack for FDT. I will not accept that without *very* good reason.